server/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -74,12 +75,18 @@ public RestRequest(NamedXContentRegistry xContentRegistry, String uri, Map<Strin
         this.xContentRegistry = xContentRegistry;
         final Map<String, String> params = new HashMap<>();
         int pathEndPos = uri.indexOf('?');
+        Exception parsingQueryStringException= null;
         if (pathEndPos < 0) {
             this.rawPath = uri;
         } else {
             this.rawPath = uri.substring(0, pathEndPos);
-            RestUtils.decodeQueryString(uri, pathEndPos + 1, params);
+            try {
+                RestUtils.decodeQueryString(uri, pathEndPos + 1, params);
+            } catch (Exception e) {
+                parsingQueryStringException = e;
+            }
         }
+        this.exception.set(parsingQueryStringException);
         this.params = params;
```

Comment:
This change still bothers me a lot. Again, I appreciate the motivation for it but the implementation that we are being forced into here leaves a lot to be desired (I am not faulting you for it, we are somewhat painted into a corner here). Another problem here is that now the object is a garbage state, we have no idea what the state of `params` is here yet we are allowing construction of this object to complete as if nothing happened. This is dangerous and I think that we should not do it.